### PR TITLE
more early stopping options (convergence and divergence threshold)

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #6795 

Adds two thresholds after which we stop training immediately (no patience). 

**Divergence threshold:** the monitor has reached a value from which we believe it cannot recover -> stop training
**Stopping threshold:** the monitor has reached a target value that is close to optimal, and we do not care about further improvement -> stop training

Now that we have multiple stopping criteria, it's best we report the reason for stopping too.

TODO:
- [x] tests for the new parameters
- [x] improve docs

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/early_stopping.py [hunk 1]
```diff
@@ -53,6 +56,10 @@ class EarlyStopping(Callback):
             monitored has stopped decreasing and in ``'max'`` mode it will stop when the quantity
             monitored has stopped increasing.
         strict: whether to crash the training if `monitor` is not found in the validation metrics.
+        check_finite: Stops training when the monitor becomes NaN or infinite. Set this argument to ``False``
+            if this behavior is undesired.
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/early_stopping.py:60]: do we want an option to turn this on/off at all?

File: pytorch_lightning/callbacks/early_stopping.py [hunk 2]
```diff
@@ -160,15 +173,48 @@ def _run_early_stopping_check(self, trainer):
         # when in dev debugging
         trainer.dev_debugger.track_early_stopping_history(self, current)
 
-        if self.monitor_op(current - self.min_delta, self.best_score):
+        should_stop, reason = self._evalute_stopping_criteria(current)
+        if should_stop:
+            self.stopped_epoch = trainer.current_epoch
+        if reason:
+            log.info(f"[{trainer.global_rank}] {reason}")
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/early_stopping.py:180]: If we log on rank zero only, and the user has `sync_dist=False` for logging, then we might not see the reason being logged because it could be rank > 0 that decided to stop.
If we log on all ranks and the user has `sync_dist=True` for logging, we will show the same message N times.

File: pytorch_lightning/callbacks/early_stopping.py [hunk 3]
```diff
@@ -53,6 +56,10 @@ class EarlyStopping(Callback):
             monitored has stopped decreasing and in ``'max'`` mode it will stop when the quantity
             monitored has stopped increasing.
         strict: whether to crash the training if `monitor` is not found in the validation metrics.
+        check_finite: Stops training when the monitor becomes NaN or infinite. Set this argument to ``False``
+            if this behavior is undesired.
+        stopping_threshold: Stop training immediately once the monitored quantity reaches this threshold.
+        divergence_threshold: Stop training as soon as the monitored quantity becomes worse than this threshold.
```
Comment (carmocca) [inline pytorch_lightning/callbacks/early_stopping.py:62]: we could use `stop_limit` and `stop_loss` to follow common financial terms

https://www.investopedia.com/articles/active-trading/091813/which-order-use-stoploss-or-stoplimit-orders.asp

File: pytorch_lightning/callbacks/early_stopping.py [hunk 4]
```diff
@@ -160,15 +173,48 @@ def _run_early_stopping_check(self, trainer):
...
+        should_stop = trainer.training_type_plugin.reduce_boolean_decision(should_stop)
+        trainer.should_stop = trainer.should_stop or should_stop
+
+    def _evalute_stopping_criteria(self, current: torch.Tensor) -> Tuple[bool, str]:
+        should_stop = False
+        reason = None
+        if self.check_finite and not torch.isfinite(current):
+            should_stop = True
+            reason = (
+                f"Monitored metric {self.monitor} = {current:.3f} is not finite."
```
Comment (carmocca) [inline pytorch_lightning/callbacks/early_stopping.py:192]: No point in formatting current as it is inf or nan

File: pytorch_lightning/callbacks/early_stopping.py [hunk 5]
```diff
@@ -160,15 +173,48 @@ def _run_early_stopping_check(self, trainer):
...
+                f" Previous best value was {self.best_score:.3f}. Signaling Trainer to stop."
+            )
+        elif self.stopping_threshold is not None and self.monitor_op(current, self.stopping_threshold):
+            should_stop = True
+            reason = (
+                f"Below tolerance {self.monitor} = {current} <= {self.stopping_threshold}"
```
Comment (carmocca) [inline pytorch_lightning/callbacks/early_stopping.py:198]:
```suggestion
                f"Below tolerance {self.monitor} = {current} <= {self.stopping_threshold}."
```

File: pytorch_lightning/callbacks/early_stopping.py [hunk 6]
```diff
@@ -160,15 +173,48 @@ def _run_early_stopping_check(self, trainer):
...
+                " Signaling Trainer to stop."
+            )
+        elif self.divergence_threshold is not None and self.monitor_op(-current, -self.divergence_threshold):
+            should_stop = True
+            reason = (
+                f"Divergence: {self.monitor} = {current} > {self.divergence_threshold}."
+                " Signaling Trainer to stop."
```
Comment (carmocca) [inline pytorch_lightning/callbacks/early_stopping.py:205]: nit: you could add "Signaling Trainer to stop" before the return instead of in every `if`

File: pytorch_lightning/callbacks/early_stopping.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/early_stopping.py:62]: @jlperla  what do you think of this name suggestion?

File: pytorch_lightning/callbacks/early_stopping.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/early_stopping.py:205]: at the end of this method before returning, I can add 
```python
        if should_stop:
            reason += " Signaling Trainer to stop."
```
there are two cases when no stopping occurs, so reason=None

File: pytorch_lightning/callbacks/early_stopping.py (hunk 3 above)
Comment (jlperla) [reply pytorch_lightning/callbacks/early_stopping.py:62]: But might not be `loss`, and most people don't know finance. These are basically optimizer settings,which is more universal. 

I think sticking with optimizer style lingo is ideal. Divergence is safe and says what it means . Normally one would call the success criteria as tolerances for optimizers. But that is because they are always comparing something (eg a value itself, changes in that value, or first order conditions) to zero.

Since this could presumably compare stopping for this other than close to zero(especially if you are tracking something where a larger number is better) , I think threshold is probably more general. But open minded of course

Comment (carmocca) [reply pytorch_lightning/callbacks/early_stopping.py:62]: Could also go super simple and go with `min_threshold` and `max_threshold`

File: pytorch_lightning/callbacks/early_stopping.py [hunk 7]
```diff
@@ -160,15 +173,48 @@ def _run_early_stopping_check(self, trainer):
...
+    def _evalute_stopping_criteria(self, current: torch.Tensor) -> Tuple[bool, str]:
+        should_stop = False
+        reason = None
+        if self.check_finite and not torch.isfinite(current):
+            should_stop = True
+            reason = (
+                f"Monitored metric {self.monitor} = {current} is not finite."
+                f" Previous best value was {self.best_score:.3f}. Signaling Trainer to stop."
+            )
+        elif self.stopping_threshold is not None and self.monitor_op(current, self.stopping_threshold):
```
Comment (carmocca) [inline pytorch_lightning/callbacks/early_stopping.py:195]: why do we use `monitor_op` here and for divergence_threshold?

Don't we just need to check with `>` and `<`?

File: pytorch_lightning/callbacks/early_stopping.py [hunk 8]
```diff
@@ -160,15 +173,48 @@ def _run_early_stopping_check(self, trainer):
...
+            should_stop = True
+            reason = (
+                f"Below tolerance {self.monitor} = {current} <= {self.stopping_threshold}."
```
Comment (carmocca) [inline pytorch_lightning/callbacks/early_stopping.py:198]: "stopping_threshold: Stop training immediately once the monitored quantity reaches this threshold."

Should this say "above or equals to tolerance" then?

**Review by carmocca — COMMENTED** [review]
Should patience and delta apply to these thresholds?

File: pytorch_lightning/callbacks/early_stopping.py (hunk 3 above)
Comment (jlperla) [reply pytorch_lightning/callbacks/early_stopping.py:62]: No because that implies a direction to them.

File: pytorch_lightning/callbacks/early_stopping.py (hunk 8 above)
Comment (jlperla) [reply pytorch_lightning/callbacks/early_stopping.py:198]: The monitor mode is needed to know which is a divergence and which is a success.

File: pytorch_lightning/callbacks/early_stopping.py (hunk 7 above)
Comment (jlperla) [reply pytorch_lightning/callbacks/early_stopping.py:195]: No.  Divergence is that it doesn't work, convergence is that it did.  They are very different in terms of how you interpret the results and map to return codes.

File: pytorch_lightning/callbacks/early_stopping.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/early_stopping.py:60]: I would say off. We will add support for occasional NaN loss training soon.

File: pytorch_lightning/callbacks/early_stopping.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/early_stopping.py]: Good idea !

File: pytorch_lightning/callbacks/early_stopping.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/early_stopping.py:62]: IMO, the name are good right now.

**Review by tchaton — APPROVED** [review]
LGTM ! Great work !

File: pytorch_lightning/callbacks/early_stopping.py (hunk 3 above)
Comment (Borda) [reply pytorch_lightning/callbacks/early_stopping.py:61]: this is not clear to me, it is you have some converging sequence so it stops when it starts diverse again? shall it be some patience for noise presence reason?
or natively it can be observing  training and validation measure and stop overfitting - when these twos tart diverse

**Review by Xl5843003 — APPROVED** [review]

File: pytorch_lightning/callbacks/early_stopping.py (hunk 3 above)
Comment (jlperla) [reply pytorch_lightning/callbacks/early_stopping.py:61]: If something is diverging, it is because you are in some sort of local minima or outside of an attractor and it could only return with some massive jumps (i.e. emulation in simulating annealing you are way off in the boons for your optima... in theory it come come back, but it might takes months.  You are better off just restarting).  So patience isn't the right thing to think of for that.

File: pytorch_lightning/callbacks/early_stopping.py [hunk 9]
```diff
@@ -160,15 +172,48 @@ def _run_early_stopping_check(self, trainer):
...
+                " Signaling Trainer to stop."
+            )
+        elif self.divergence_threshold is not None and self.monitor_op(-current, -self.divergence_threshold):
+            should_stop = True
+            reason = (
+                f"Divergence: {self.monitor} = {current} > {self.divergence_threshold}."
```
Comment (jlperla) [inline pytorch_lightning/callbacks/early_stopping.py:203]: Minor point:  whether it is
```
f"Divergence: {self.monitor} = {current} > {self.divergence_threshold}."
```
or
```
f"Divergence: {self.monitor} = {current} > {self.divergence_threshold}."
```
should depend on the `monitor_op`, right?  Maybe have a `f"Divergence: {self.monitor} = {current} {op_to_string(self.monitor_op)} {self.divergence_threshold}."`
or something like that, where you fill `op_to_string` with whatever you need to turn it into a `>`, etc.?

Similarly, I think you could do the same with the  "successful" convergence below/above the target.
```python
 f"Below tolerance {self.monitor} = {current} {op_to_string(self.monitor_op)} {self.stopping_threshold}  ."
```
where you would have to ensure the order is correct as it is going form the other direction.

As I said though, minor.

**Comment by jlperla** [comment]
@tchaton @awaelchli This all looks great to me.  I put in one minor comment about the "reason" strings only being correct for one of the monitor_ops, but I also think that could wait and do it as a separate issue later.  I personally am unlikely to use the other direction for the monitor_op anytime soon.

**Review by kaushikb11 — APPROVED** [review]
Love it!

**Review by carmocca — APPROVED** [review]
